packages/apollo-federation/src/composition/validate/postComposition/externalUnused.ts

```diff
@@ -15,108 +18,183 @@ import {
+           * referenced by a selection set inside the @requires of User.isAdmin
+           *
+           *    extend type User @key(fields: "id") {
+           *      roles: AccountRoles!
+           *      isAdmin: Boolean! @requires(fields: "roles { canWrite permission { status } }")
+           *    }
+           *    extend type AccountRoles {
+           *      canWrite: Boolean @external
+           *      permission: Permission @external
+           *    }
+           *
+           *    extend type Permission {
+           *      status: String @external
+           *    }
+           *
+           * So, we need to search for fields with requires, then parse the selection sets,
+           * and try to recursively find the external field's PARENT type, then the external field's name
+           */
+          const hasMatchingRequiresOnAnotherType = Object.values(
+            schema.getTypeMap(),
+          ).some(namedType => {
+            if (!isObjectType(namedType)) return false;
+            // for every object type, loop over its fields and find fields
+            // with requires directives
+            return Object.values(namedType.getFields()).some(field => {
+              const directivesOnField = findDirectivesOnTypeOrField(
+                field.astNode,
                 'requires',
               );
 
-              return (
-                hasMatchingFieldInDirectives({
-                  directives: providesDirectives,
-                  fieldNameToMatch: externalFieldName,
-                  namedType,
-                }) ||
-                hasMatchingFieldInDirectives({
-                  directives: requiresDirectives,
-                  fieldNameToMatch: externalFieldName,
-                  namedType,
-                })
-              );
-            },
-          );
+              return directivesOnField.some(directive => {
+                if (!directive.arguments) return false;
+                const selections =
+                  isStringValueNode(directive.arguments[0].value) &&
+                  parseSelections(directive.arguments[0].value.value);
 
-          if (!(hasMatchingKeyOnType || hasMatchingProvidesOrRequires)) {
-            errors.push(
-              errorWithCode(
-                'EXTERNAL_UNUSED',
-                logServiceAndType(serviceName, typeName, externalFieldName) +
-                  `is marked as @external but is not used by a @requires, @key, or @provides directive.`,
-              ),
+                if (!selections) return false;
+                return selectionIncludesField({
+                  selections,
+                  selectionSetType: namedType,
+                  typeToFind: parentType,
+                  fieldToFind: externalFieldName,
+                });
+              });
+            });
+          });
+
+          if (hasMatchingRequiresOnAnotherType) continue;
+
+          const hasMatchingProvidesOrRequiresOnType = Object.values(
```

Comment:
Is this block still needed? The `xxxOnAnotherType` blocks are looking at the whole schema already, so I think they eliminate the need for this loop.

Reply:
😳 I think you just may be right! That's huge

Reply:
Ah, so close. The `provides` work was an overlap with the original code, so I was able to remove some of that, but the `requires` isn't a direct overlap with the new stuff, so that had to stay in 💀